crates/trie/parallel/src/proof_task.rs

```diff
@@ -0,0 +1,319 @@
+        for item in &mut proof_task_txs {
+            let provider_ro = view.provider_ro()?;
+            let tx = provider_ro.into_tx();
+            *item = ProofTaskTx::new(tx, task_ctx.clone());
+        }
+
+        Ok(Self {
+            pending_proofs: VecDeque::new(),
+            executor,
+            proof_task_txs,
+            running_proof_tasks: Vec::new(),
+            proof_task_rx,
+        })
+    }
+}
+
+impl<Tx> ProofTaskManager<Tx>
+where
+    Tx: DbTx + 'static,
+{
+    /// Spawns the proof task on the executor, with the input multiproof targets.
+    ///
+    /// If a task cannot be spawned immediately, this will be queued for completion later.
+    pub fn queue_proof_task(
+        &mut self,
+        input: StorageProofInput,
+        sender: Sender<StorageProofResult>,
+    ) {
+        self.pending_proofs.push_back((input, sender));
+    }
+
+    /// Spawns the next queued proof task on the executor with the given input, if there are any
+    /// transactions available.
+    pub fn try_spawn_next(&mut self) {
+        let Some(proof_task_tx) = self.proof_task_txs.pop() else { return };
+
+        let Some((pending_proof, sender)) = self.pending_proofs.pop_front() else {
+            // if there are no targets to do anything with put the tx back
+            self.proof_task_txs.push(proof_task_tx);
+            return
+        };
+
+        let (tx, rx) = channel();
+        self.executor.spawn_blocking(move || {
+            let output = proof_task_tx.storage_proof(pending_proof);
+            let _ = tx.send(output);
+        });
+        self.running_proof_tasks.push(RunningProofTask::new(rx, sender));
+    }
+
+    /// Loops, managing the proof tasks, and sending new tasks to the executor.
+    pub fn run(mut self) {
+        loop {
+            // TODO: condvar for
+            // * either running or proof tasks have stuff
+            // * otherwise yield thread
+            let message = match self.proof_task_rx.try_recv() {
+                Ok(message) => match message {
+                    ProofTaskMessage::StorageProof(input) => Some(input),
+                    ProofTaskMessage::Terminate => return,
+                },
+                Err(TryRecvError::Empty) => None,
+                Err(TryRecvError::Disconnected) => return,
+            };
+
+            if let Some((input, sender)) = message {
+                self.queue_proof_task(input, sender);
+            }
+
+            // try spawning the next task
+            self.try_spawn_next();
+
+            // tracks the disconnected senders
+            let mut tasks_to_remove = Vec::new();
+
+            // poll each task
+            for (idx, running_task) in self.running_proof_tasks.iter().enumerate() {
+                let RunningProofTask { task_receiver, sender } = running_task;
+                match task_receiver.try_recv() {
+                    Ok(ProofTaskOutput { tx, result }) => {
+                        let _ = sender.send(result);
```

Comment:
this feels a bit weird, why are we awaiting both the result and the sender here only to send the result through that sender

Reply:
I guess we could send the result directly in the spawned storage proof task, and only pass the tx back in the `ProofTaskOutput`

Reply:
now we send the result directly to the sender in `storage_proof`